### PR TITLE
fix: [CDS-82778]: Migrate to versions endpoint instead of tags endpiont for GAR

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
Earlier we used the tags endpoint to fetch GAR Artifact tags which did not have information about the timestamps, now we will use the versions endpoint which will contain the information about timestamps.

I tested that the API response is successful and we are listing the API according to the time of creation and lexically if the the timestamp is same.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55094)

### CODE REVIEW

File: 960-api-services/src/main/java/io/harness/artifacts/comparator/BuildDetailsInternalTimeComparator.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.Comparator;
+
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = true, components = {HarnessModuleComponent.CDS_ARTIFACTS})
+@OwnedBy(HarnessTeam.CDC)
+public class BuildDetailsInternalTimeComparator implements Comparator<BuildDetailsInternal> {
+  @Override
+  public int compare(BuildDetailsInternal o1, BuildDetailsInternal o2) {
+    if (o1.getImagePushedAt() == null || o2.getImagePushedAt() == null
+        || o1.getImagePushedAt().compareTo(o2.getImagePushedAt()) == 0) {
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/comparator/BuildDetailsInternalTimeComparator.java:25]: Should we opt o1 if o1.getImagePushedAt() != null and o2.getImagePushedAt() == null?

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 2]
```diff
@@ -250,37 +253,43 @@ private boolean isSuccessful(int code, String errormessage) {
...
-                  -> StringUtils.isBlank(versionRegex) || new RegexFunctor().match(versionRegex, build.getNumber()))
-              .collect(toList());
 
-      return buildDetails.stream().sorted(new BuildDetailsInternalComparatorDescending()).collect(toList());
+  public List<BuildDetailsInternal> processPage(
+      GarPackageVersionResponse versionsPage, String versionRegex, GarInternalConfig garinternalConfig) {
+    if (versionsPage != null && EmptyPredicate.isNotEmpty(versionsPage.getVersions())) {
+      List<BuildDetailsInternal> buildDetails = new ArrayList<>();
+      for (GarVersions garVersions : versionsPage.getVersions()) {
+        buildDetails.addAll(
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:262]: Please check if there is an image without tags, do we get null tag for it?

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 3]
```diff
@@ -40,8 +41,10 @@
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
 import java.io.IOException;
+import java.time.Instant;
 import java.util.ArrayList;
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:45]: Please add UTs

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55094)* status: 9 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/artifacts/comparator/BuildDetailsInternalTimeComparator.java` line 25 at r1](https://reviewable.io/reviews/harness/harness-core/55094#gh-1387577911:-NinnR-R5IQdPDXmvRfD:b-a3pjyz) ([raw file](https://github.com/harness/harness-core/blob/fb2f9a247fcb392b75f93a6ce723b811fcc1b4ae/960-api-services/src/main/java/io/harness/artifacts/comparator/BuildDetailsInternalTimeComparator.java#L25)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

Should we opt o1 if o1.getImagePushedAt() != null and o2.getImagePushedAt() == null?
</blockquote></details>

If image pushed at is not set for any of the fields then we have to compare by Number.
___
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 45 at r1](https://reviewable.io/reviews/harness/harness-core/55094#gh-1387586576:-Nino0iz9kyPKJxKDp7A:bxlogbl) ([raw file](https://github.com/harness/harness-core/blob/fb2f9a247fcb392b75f93a6ce723b811fcc1b4ae/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L45)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

Please add UTs
</blockquote></details>

I have updated the previous UT with my changes
___
*[`960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java` line 262 at r1](https://reviewable.io/reviews/harness/harness-core/55094#gh-1387584080:-NinnqaQCLZ7rZ-m2Bkb:b32jxxi) ([raw file](https://github.com/harness/harness-core/blob/fb2f9a247fcb392b75f93a6ce723b811fcc1b4ae/960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java#L262)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

Please check if there is an image without tags, do we get null tag for it?
</blockquote></details>

We get null for relatedTags, handled that case

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 3 above)
Comment (sarthakkasat) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:45]: Done

**Review by puthrayaharness — APPROVED** [review]

**Review by Abhishek-CDC — DISMISSED** [review]

**Comment by Abhishek-CDC** [comment]
trigger uts

**Comment by Abhishek-CDC** [comment]
Please wait for UTs to run before merging

**Comment by sarthakkasat** [comment]
trigger uts

**Comment by sarthakkasat** [comment]
trigger messagecheck

**Review by Abhishek-CDC — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]

**Comment by sarthakkasat** [comment]
trigger review

**Pull request merged**